presto/utils.py

```diff
@@ -216,88 +180,111 @@ def process_parquet(df: pd.DataFrame) -> pd.DataFrame:
-    new_start_date = df.groupby(["sample_id"])["timestamp"].min()
-    df["start_date"] = df["sample_id"].map(new_start_date)
 
-    # Also reassign end_date to the maximum timestamp
-    new_end_date = df.groupby(["sample_id"])["timestamp"].max()
-    df["end_date"] = df["sample_id"].map(new_end_date)
-
-    # reinitialize timestep_ind
-    df["timestamp_ind"] = (df["timestamp"].dt.year * 12 + df["timestamp"].dt.month) - (
-        df["start_date"].dt.year * 12 + df["start_date"].dt.month
+        if len(samples_after_end_date) > 0 or len(samples_before_start_date) > 0:
+            logger.warning(
+                f"""\
+    Removing {len(samples_after_end_date)} \
+    samples with valid_time after the end_date \
+    and {len(samples_before_start_date)} samples with valid_time before the start_date"""
+            )
+            df = df[~df["sample_id"].isin(samples_before_start_date)]
+            df = df[~df["sample_id"].isin(samples_after_end_date)]
+
+        # add timesteps before the start_date where needed
+        intermediate_dummy_df = pd.DataFrame()
+        for n_ts_to_add in range(1, MIN_EDGE_BUFFER + 1):
+            samples_to_add_ts_before_start = latest_obs_position[
+                (MIN_EDGE_BUFFER - latest_obs_position["valid_position"]) >= -n_ts_to_add
+            ].index
+            dummy_df = df[
+                (df["sample_id"].isin(samples_to_add_ts_before_start)) & (df["timestamp_ind"] == 0)
+            ].copy()
+            dummy_df["timestamp"] = dummy_df["timestamp"] - pd.DateOffset(
+                months=n_ts_to_add
+            )  # type: ignore
+            dummy_df[feature_columns] = NODATAVALUE
+            intermediate_dummy_df = pd.concat([intermediate_dummy_df, dummy_df])
+        df = pd.concat([df, intermediate_dummy_df])
+
+        # add timesteps after the end_date where needed
+        intermediate_dummy_df = pd.DataFrame()
+        for n_ts_to_add in range(1, MIN_EDGE_BUFFER + 1):
+            samples_to_add_ts_after_end = latest_obs_position[
+                (MIN_EDGE_BUFFER - latest_obs_position["valid_position_diff"]) >= n_ts_to_add
+            ].index
+            dummy_df = df[
+                (df["sample_id"].isin(samples_to_add_ts_after_end)) & (df["is_last_available_ts"])
+            ].copy()
+            dummy_df["timestamp"] = dummy_df["timestamp"] + pd.DateOffset(
+                months=n_ts_to_add
+            )  # type: ignore
+            dummy_df[feature_columns] = NODATAVALUE
+            intermediate_dummy_df = pd.concat([intermediate_dummy_df, dummy_df])
+        df = pd.concat([df, intermediate_dummy_df])
+
+    # create timestep_ind
+    df["timestamp_ind"] = df.groupby("sample_id")["timestamp"].rank().astype(int)
+    df["available_timesteps"] = df["sample_id"].map(
+        df.groupby("sample_id")["timestamp"].nunique().astype(int)
     )
+    index_columns.append("available_timesteps")
 
     # check for missing timestamps in the middle of timeseries
```

Comment:
Does this comment belong to the actual code on the next line (being the concatenation of dataframes)?